File: src/TodoDetails.ts
import {Change, Chunk} from "parse-diff";
import {argumentContext} from "./ArgumentContext";
import {assignFlow, escapeForRegExp, lineBreak} from "./helpers";
import {prNr} from "./RepoContext";
import {getUsername} from "./GitHubContext";
import {generateAssignedTo} from "./AssignHelper";


/**
 * Get the file boundaries of the hunk
 */
export function getFileBoundaries(changes: any[], line: number, paddingTop = 0, paddingBottom = 2) {

    let firstChange = changes[0]
    let lastChange = changes[changes.length - 1]

    const firstChangedLine = firstChange.ln || firstChange.ln2;
    const lastChangedLine = lastChange.ln || lastChange.ln2;

    let start;
    if (line == firstChangedLine) {
        start = Math.max(1, line - paddingTop);
    } else {
        start = Math.max(line - paddingTop, firstChangedLine)
    }

    // we dont know the actual lines count of the file, so we cant add something like minPadding (at least for the bottom)
    let end = Math.min(line + paddingBottom, lastChangedLine)

    return {start, end}
}

/**
 * Prepares some details about the TO_DO
 */
export function checkForBody(changes: Change[], changeIndex: number, beforeTag: string) {
    const bodyPieces = []
    const nextChanges = changes.slice(changeIndex + 1)

    const BODY_REG = new RegExp(`${escapeForRegExp(beforeTag)}\\W*(?<keyword>${argumentContext.bodyKeywords.join('|')})\\b\\W*(?<body>.*)`, !argumentContext.caseSensitive ? 'i' : '')

    for (const change of nextChanges) {
        const matches = BODY_REG.exec(change.content)

        if (!matches) break

        if (!matches.groups?.body) {
            bodyPieces.push('\n')
        } else {
            if (bodyPieces.length > 0 && bodyPieces[bodyPieces.length - 1] !== '\n') bodyPieces.push(' ')
            bodyPieces.push(lineBreak(matches.groups.body).trim())
        }
    }

    return bodyPieces.length ? bodyPieces.join('') : false
}

export function splitTagsFromTitle(title: string): string[] {

    if (!title)
        return [];

    const getMatch = () => {
        return title.match(new RegExp(`\\[([^\\]]+)]$`));
    }

    const tags: string[] = [];

    let match = getMatch();

    while (match) {
        tags.push(match[1].trim());
        title = title.replace(match[0], "").trimEnd();
        match = getMatch();
    }

    return tags;
}

export function getDetails(chunk: Chunk, line: number) {

    const username = getUsername()

    // Generate a string that expresses who the issue is assigned to
    const assignedToString = generateAssignedTo(username, prNr)

    const assignees = assignFlow(username);

    let range: false | string
    if (!argumentContext.blobLines) {
        // Don't show the blob
        range = false
    } else {
        const {
            start,
            end
        } = getFileBoundaries(chunk.changes, line, argumentContext.blobLinesBefore, argumentContext.blobLines)
        range = start === end ? `L${start}` : `L${start}-L${end}`
    }

    return {
        username,
        assignedToString,
        number: prNr,
        range,
        assignees
    }
}

File: src/TodoIssue.Core/TodoDetails.cs
using System.Text.RegularExpressions;

namespace TodoIssue.Core;

public static partial class TodoDetails
{
    public static (int Start, int End) GetFileBoundaries(IReadOnlyList<DiffChange> changes, int line, int paddingTop = 0, int paddingBottom = 2)
    {
        var firstChangedLine = changes[0].Line;
        var lastChangedLine = changes[^1].Line;
        var start = line == firstChangedLine ? Math.Max(1, line - paddingTop) : Math.Max(line - paddingTop, firstChangedLine);
        var end = Math.Min(line + paddingBottom, lastChangedLine);
        return (start, end);
    }

    public static string? CheckForBody(AppOptions options, IReadOnlyList<DiffChange> changes, int changeIndex, string beforeTag)
    {
        var body = new List<string>();

        foreach (var change in changes.Skip(changeIndex + 1))
        {
            if (!TryReadBodyLine(options, change.Content, beforeTag, out var line))
            {
                break;
            }

            if (string.IsNullOrEmpty(line))
            {
                body.Add("\n");
                continue;
            }

            if (body.Count > 0 && body[^1] != "\n")
            {
                body.Add(" ");
            }

            body.Add(Helpers.LineBreak(line).Trim());
        }

        return body.Count == 0 ? null : string.Concat(body);
    }

    public static (string Title, IReadOnlyList<string> Tags) SplitTagsFromTitle(string title)
    {
        var tags = new List<string>();

        while (true)
        {
            var match = TagRegex().Match(title);
            if (!match.Success)
            {
                return (title, tags);
            }

            tags.Add(match.Groups[1].Value.Trim());
            title = TagRegex().Replace(title, "").TrimEnd();
        }
    }

    public static (string Content, IReadOnlyList<string> Assignees) GetMentionedAssignees(string content, bool clipMentionedFromContent)
    {
        var matches = MentionRegex().Matches(content);
        var assignees = matches.Select(match => Helpers.StripAt(match.Value)).Distinct(StringComparer.Ordinal).ToList();
        return (clipMentionedFromContent ? MentionRegex().Replace(content, "").Trim() : content, assignees);
    }

    public static string GenerateAssignedTo(AppOptions options, TodoRunContext context, IReadOnlyList<string> assignees)
    {
        if (assignees.Count == 0)
        {
            return "";
        }

        var assigner = Helpers.ReduceToList(assignees.Select(Helpers.AddAt));
        var customUsers = options.AutoAssignUsers?.Select(Helpers.StripAt).Order(StringComparer.Ordinal).ToArray();
        var sortedAssignees = assignees.Order(StringComparer.Ordinal).ToArray();

        if (customUsers is not null && customUsers.SequenceEqual(sortedAssignees))
        {
            return context.IssueNumber is not null ? $" cc {assigner}" : $" It's been automagically assigned to {assigner}.";
        }

        if (options.AutoAssignEnabled && context.Username is not null && assignees.SequenceEqual([context.Username]))
        {
            return context.IssueNumber is not null ? $" cc @{assigner}." : $" It's been assigned to {assigner} because they committed the code.";
        }

        return context.IssueNumber is not null ? $" cc {assigner}" : $" It's been assigned to {assigner} because they were mentioned in the comment.";
    }

    private static bool TryReadBodyLine(AppOptions options, string content, string beforeTag, out string body)
    {
        body = "";

        if (!content.StartsWith(beforeTag, options.CaseSensitive ? StringComparison.Ordinal : StringComparison.OrdinalIgnoreCase))
        {
            return false;
        }

        var rest = content[beforeTag.Length..];
        rest = TrimNonWordStart(rest);

        if (options.BodyKeywords.Count == 0)
        {
            body = rest;
            return true;
        }

        var comparison = options.CaseSensitive ? StringComparison.Ordinal : StringComparison.OrdinalIgnoreCase;
        foreach (var keyword in options.BodyKeywords)
        {
            if (!rest.StartsWith(keyword, comparison))
            {
                continue;
            }

            var afterKeyword = rest[keyword.Length..];
            if (afterKeyword.Length > 0 && IsWord(afterKeyword[0]))
            {
                continue;
            }

            body = TrimNonWordStart(afterKeyword);
            return true;
        }

        return false;
    }

    private static string TrimNonWordStart(string value)
    {
        var index = 0;
        while (index < value.Length && !IsWord(value[index]))
        {
            index++;
        }

        return value[index..];
    }

    private static bool IsWord(char c) => char.IsLetterOrDigit(c) || c == '_';

    [GeneratedRegex(@"\[([^\]]+)]$")]
    private static partial Regex TagRegex();

    [GeneratedRegex(@"@[a-zA-Z\d@._-]+\b")]
    private static partial Regex MentionRegex();
}
